geohisto/models.py

```diff
@@ -167,57 +227,127 @@ def add_modification(self, modification):
-
-    def get_ancestors(self, towns):
-        """Iterator across ancestors of the current town."""
-        for town in towns.with_successors():
-            for successor_id in town.successors.split(';'):
-                if successor_id == self.id:
-                    yield town
+Record = namedtuple('Record', [
+    'depcom', 'mod', 'eff', 'nccoff', 'nccanc', 'comech', 'dep', 'com',
+    'depanc', 'last', 'effdate'
+])
 
-    def add_successor(self, successor):
-        """Append the given successor to the current list if any."""
-        if self.successors:
-            successor = self.successors + ';' + successor
-        return self._replace(**{'successors': successor})
 
-    def replace_successor(self, old_successor, new_successor):
-        """Replace a successor within the current list."""
-        successors = self.successors.replace(old_successor, new_successor)
-        return self._replace(**{'successors': successors.strip(';')})
+class Intercommunalities(Collection, defaultdict):
+    def latest(self, siren):
+        """Get the latest valid intercommunality for a given `siren`."""
+        _items = self.filter(siren=siren, end_date=END_DATE)
+        # _items.sort(key=lambda item: item.end_date, reverse=True)
+        return _items[0]
 
-    def remove_successor(self, successor):
-        """Remove the given successor from the current list."""
-        successors = ';'.join([
-            succ for succ in self.successors.split(';')
-            if succ != successor])
-        return self._replace(**{'successors': successors})
+    def valid_at(self, valid_date, siren=None):
+        """
+        Return a list of Intercommunalities existing at the given `valid_date`.
+        """
+        # Beware, ternary operator is tricky here, keep it explicit.
+        if siren:
+            _items = self.filter(siren=siren)
+        else:
+            _items = self.values()
+        return [item for item in _items if item.valid_at(valid_date)]
+    #
+    # def get_current(self, siren, valid_date):
+    #     """
+    #     Try to return the more pertinent Intercommunality
+    #     given a siren and date.
+    #     """
+    #     try:
+    #         return self.valid_at(valid_date, siren=siren)[0]
+    #     except IndexError:
+    #         return self.latest(siren)
 
-    def clear_successors(self):
-        """Remove all successors."""
-        return self._replace(**{'successors': ''})
+    @property
+    def open_sirens(self):
+        return set(item.siren for item in self.filter(end_date=END_DATE))
+
+    def ends(self, siren, year, reason):
+        intercommunality = self.latest(siren)
+        self.upsert(intercommunality.end_on(year, reason))
+
+    def update(self, intercommunality, year):
+        current = self.latest(intercommunality.siren)
+        changes = []
+        if intercommunality.towns != current.towns:
+            # This is a perimeter change
+            changes.append(END_INNER)
```

Comment:
`INTERCOMMUNALITY_PERIMETER_CHANGE` does not fit, it can simply be a component town renamed, merged or splitted... perimeter does not change is this case.
It took `INTERCOMMUNALITY_INNER_CHANGE`